consensus/fork_choice/src/fork_choice.rs

```diff
@@ -1499,6 +1532,112 @@ where
     }
 }
 
+fn process_justification_and_finalization_from_progressive_cache<E, T>(
+    state: &BeaconState<E>,
+    spec: &ChainSpec,
+    progressive_balances_mode: ProgressiveBalancesMode,
+    log: &Logger,
+) -> Result<JustificationAndFinalizationState<E>, Error<T::Error>>
+where
+    E: EthSpec,
+    T: ForkChoiceStore<E>,
+{
+    let justification_and_finalization_state = JustificationAndFinalizationState::new(state);
+    if state.current_epoch() <= E::genesis_epoch() + 1 {
+        return Ok(justification_and_finalization_state);
+    }
+
+    // Load cached balances
+    let progressive_balances_cache: &ProgressiveBalancesCache = state.progressive_balances_cache();
+    let previous_target_balance =
+        progressive_balances_cache.previous_epoch_target_attesting_balance()?;
```

Comment:
> Perhaps it would be more in the spirit of Checked mode for us just to default back to the non-optimized method if we happen to discover an uninitialized progressive_balances_cache?

This is a great point! `Checked` mode should just always fallback to the non-optimized method if there's anything wrong with the `progressive_balances_cache`. I've re-arranged the code a bit so that it falls back to the original method if this function failed for whatever reason in `Checked` mode.

Updated in https://github.com/sigp/lighthouse/pull/4362/commits/72935e14189d34d814ad777bd528cbc4412594b0

Reply:
Oops, missed a check in the previous commit and broke some tests. Fixed in https://github.com/sigp/lighthouse/pull/4362/commits/f49daa08f9bb7b053a3684dffc5143310da04d4e.